### DISCUSSION TITLE
Request support for mise

### DISCUSSION THREAD

**Discussion by rakuyoMo** [fastlane/fastlane#21955] [General] [answered]
[mise](https://github.com/jdx/mise) was formerly called rtx. It is an excellent CLI tool. Currently, fastlane does not support downloading and installation through it.

My commonly used tools such as [SwiftLint](https://github.com/klundberg/asdf-swiftlint.git), [SwiftFormat](https://github.com/younke/asdf-swiftformat.git), [SwiftGen](https://github.com/younke/asdf-swiftgen.git), [CocoaPods](https://github.com/ronnnnn/asdf-cocoapods.git), etc. are already supported, and currently only fastlane is missing. I hope this project can consider supporting mise, thank you!

**Comment by Lontronix** [comment]
would love to see this happen as well

**Comment by rakuyoMo** [comment]
@KrauseFx @crazymanish @rogerluan @joshdholtz 
Hello great contributors, could you please pay attention to this issue? Thank you very much!

**Comment by mollyIV** [comment, marked as answer]
Hi, 

I've been working on Fastlane support for Mise lately. You can check out the repository here: 
* https://github.com/mollyIV/asdf-fastlane.

The implementation is heavily inspired by the Cocoapods Mise plugin: 
* https://github.com/ronnnnn/asdf-cocoapods.

If you'd like to try it out, you can modify your `.mise.toml` config file as follows:

```yaml
[tools]
fastlane = '2.221.0'
ruby = '3.0'

[plugins]
fastlane = 'git@github.com:mollyIV/asdf-fastlane.git'
```

Unfortunately, I haven't been able to get it to work and I'm getting this error:

```bash
<internal:/Users/daniel/.local/share/mise/installs/ruby/3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require': cannot load such file -- fastlane/cli_tools_distributor (LoadError)
        from <internal:/Users/daniel/.local/share/mise/installs/ruby/3.2.5/lib/ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:86:in `require'
        from /Users/daniel/.local/share/mise/installs/fastlane/2.221.0/bin/fastlane:12:in `<main>'
```

Any help would be appreciated. Thanks!

**Reply by rakuyoMo** [reply]
Hello, first of all, thank you for paying attention to this issue and trying it.

I used the following configuration to try to install fastlane through mise:

```toml
[plugins]
# Although it points to my forked repository, I haven't made any changes
fastlane = 'https://github.com/rakuyoMo/asdf-fastlane.git'

[tools]
...
fastlane = '2.222.0'
ruby = '3.3.0'
```

Both fastlane and ruby ​​are installed correctly, but when I execute `mise exec -- fastlane --version`, an error is reported:

```shell
Ignoring bigdecimal-3.1.8 because its extensions are not built. Try: gem pristine bigdecimal --version 3.1.8
Ignoring bigdecimal-3.1.6 because its extensions are not built. Try: gem pristine bigdecimal --version 3.1.6
Ignoring bigdecimal-3.1.5 because its extensions are not built. Try: gem pristine bigdecimal --version 3.1.5
Ignoring bigdecimal-3.1.4 because its extensions are not built. Try: gem pristine bigdecimal --version 3.1.4
Ignoring commonmarker-0.23.7 because its extensions are not built. Try: gem pristine commonmarker --version 0.23.7
Ignoring digest-crc-0.6.5 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.5
Ignoring digest-crc-0.6.4 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.4
Ignoring digest-crc-0.6.3 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.3
Ignoring eventmachine-1.2.7 because its extensions are not built. Try: gem pristine eventmachine --version 1.2.7
Ignoring executable-hooks-1.6.1 because its extensions are not built. Try: gem pristine executable-hooks --version 1.6.1
fastlane installation at path:
/Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/bin/fastlane
-----------------------------
Ignoring json-2.7.2 because its extensions are not built. Try: gem pristine json --version 2.7.2
Ignoring json-2.7.1 because its extensions are not built. Try: gem pristine json --version 2.7.1
Ignoring json-2.6.3 because its extensions are not built. Try: gem pristine json --version 2.6.3
Ignoring json-2.6.2 because its extensions are not built. Try: gem pristine json --version 2.6.2
Ignoring json-2.6.1 because its extensions are not built. Try: gem pristine json --version 2.6.1
Ignoring json-2.5.1 because its extensions are not built. Try: gem pristine json --version 2.5.1
Ignoring strscan-3.1.0 because its extensions are not built. Try: gem pristine strscan --version 3.1.0
Ignoring nkf-0.2.0 because its extensions are not built. Try: gem pristine nkf --version 0.2.0
Ignoring unf_ext-0.0.8.2 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.8.2
Ignoring unf_ext-0.0.8.1 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.8.1
Ignoring unf_ext-0.0.8 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.8
Ignoring unf_ext-0.0.7.7 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.7.7
Ignoring digest-crc-0.6.5 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.5
Ignoring digest-crc-0.6.4 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.4
Ignoring digest-crc-0.6.3 because its extensions are not built. Try: gem pristine digest-crc --version 0.6.3
[⠸] 🚀 Ignoring ffi-1.16.3 because its extensions are not built. Try: gem pristine ffi --version 1.16.3
Ignoring ffi-1.16.2 because its extensions are not built. Try: gem pristine ffi --version 1.16.2
Ignoring ffi-1.16.1 because its extensions are not built. Try: gem pristine ffi --version 1.16.1
Ignoring ffi-1.15.5 because its extensions are not built. Try: gem pristine ffi --version 1.15.5
Ignoring ffi-1.15.4 because its extensions are not built. Try: gem pristine ffi --version 1.15.4
Ignoring ffi-1.15.3 because its extensions are not built. Try: gem pristine ffi --version 1.15.3
Ignoring gem-wrappers-1.4.0 because its extensions are not built. Try: gem pristine gem-wrappers --version 1.4.0
Ignoring google-protobuf-3.21.12 because its extensions are not built. Try: gem pristine google-protobuf --version 3.21.12
Ignoring http_parser.rb-0.8.0 because its extensions are not built. Try: gem pristine http_parser.rb --version 0.8.0
Ignoring json-2.7.2 because its extensions are not built. Try: gem pristine json --version 2.7.2
Ignoring json-2.7.1 because its extensions are not built. Try: gem pristine json --version 2.7.1
Ignoring json-2.6.3 because its extensions are not built. Try: gem pristine json --version 2.6.3
Ignoring json-2.6.2 because its extensions are not built. Try: gem pristine json --version 2.6.2
Ignoring json-2.6.1 because its extensions are not built. Try: gem pristine json --version 2.6.1
Ignoring json-2.5.1 because its extensions are not built. Try: gem pristine json --version 2.5.1
Ignoring nio4r-2.5.4 because its extensions are not built. Try: gem pristine nio4r --version 2.5.4
[⠼] 🚀 Ignoring nkf-0.2.0 because its extensions are not built. Try: gem pristine nkf --version 0.2.0
Ignoring racc-1.6.2 because its extensions are not built. Try: gem pristine racc --version 1.6.2
Ignoring racc-1.5.2 because its extensions are not built. Try: gem pristine racc --version 1.5.2
Ignoring strscan-3.1.0 because its extensions are not built. Try: gem pristine strscan --version 3.1.0
Ignoring unf_ext-0.0.8.2 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.8.2
Ignoring unf_ext-0.0.8.1 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.8.1
Ignoring unf_ext-0.0.8 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.8
Ignoring unf_ext-0.0.7.7 because its extensions are not built. Try: gem pristine unf_ext --version 0.0.7.7
Ignoring websocket-driver-0.7.3 because its extensions are not built. Try: gem pristine websocket-driver --version 0.7.3
[⠹] 🚀 /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane/lib/fastlane/cli_tools_distributor.rb:126:in `ensure in take_off': uninitialized constant FastlaneCore::UpdateChecker (NameError)

        FastlaneCore::UpdateChecker.show_update_status('fastlane', Fastlane::VERSION)
                    ^^^^^^^^^^^^^^^
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane/lib/fastlane/cli_tools_distributor.rb:126:in `take_off'
	from /Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/bin/fastlane:23:in `<main>'
<internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require': linked to incompatible /Users/rakuyo/.rvm/rubies/ruby-2.7.4/lib/libruby.2.7.dylib - /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/nkf-0.2.0/lib/nkf.bundle (LoadError)
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/nkf-0.2.0/lib/nkf.rb:5:in `<top (required)>'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/nkf-0.2.0/lib/kconv.rb:13:in `<top (required)>'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/CFPropertyList-3.0.7/lib/cfpropertylist/rbCFPropertyList.rb:3:in `<top (required)>'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/CFPropertyList-3.0.7/lib/cfpropertylist.rb:3:in `<top (required)>'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane_core/lib/fastlane_core/core_ext/cfpropertylist.rb:4:in `<top (required)>'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane_core/lib/fastlane_core/ipa_file_analyser.rb:4:in `require_relative'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane_core/lib/fastlane_core/ipa_file_analyser.rb:4:in `<top (required)>'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane_core/lib/fastlane_core.rb:23:in `require_relative'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane_core/lib/fastlane_core.rb:23:in `<top (required)>'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane/lib/fastlane.rb:1:in `<top (required)>'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from <internal:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/lib/ruby/3.3.0/rubygems/core_ext/kernel_require.rb>:127:in `require'
	from /Users/rakuyo/.rvm/gems/ruby-2.7.4/gems/fastlane-2.216.0/fastlane/lib/fastlane/cli_tools_distributor.rb:41:in `take_off'
	from /Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/bin/fastlane:23:in `<main>'
```

There are some prerequisites: I have also installed ruby ​​and fastlane in other locations on my computer, and the versions are different from those in the above mise configuration file. This can also be seen from the above error log.

I think your mise plugin may be ok, because it can download and install fastlane correctly.

Could the problem be with fastlane itself? It can't find the corresponding ruby ​​version correctly. Or maybe you need to pass the path of the ruby ​​installed by mise to fastlane in the mise plugin?

**Reply by rakuyoMo** [reply]
This may also be a mise issue. I searched for this: https://github.com/jdx/mise/issues/1858

**Reply by mollyIV** [reply]
After talking with the very kind folks from [Tuist](https://tuist.io/) team, it appears that to resolve the error, we need to add the following to the `bin/fastlane` file (located at `~/.local/share/mise/installs/fastlane/<version>/bin/fastlane`): 

```ruby
...

ENV['BUNDLE_GEMFILE'] = File.expand_path('../../Gemfile', __FILE__) # 🆕
require "bundler/setup" # 🆕

require "fastlane/cli_tools_distributor"

...
```
The modification seems to instruct the Bundler on where the dependencies are defined.

[CocoaPods](https://github.com/CocoaPods/CocoaPods/blob/master/bin/pod#L24-L32) uses a similar approach. 

There's one caveat: when you change Ruby version, you need to reinstall the Fastlane plugin: 

```bash
$ mise plugins uninstall fastlane -p
$ mise install 
```

Now, we need to consider the implications of this change to the `bin/fastlane`. Any ideas?

**Reply by rakuyoMo** [reply]
Wow, I also use tuist, I didn't expect they knew about this issue. 
mise has some [built-in environment variables](https://mise.jdx.dev/configuration.html#environment-variables), but I don't think it's a good idea to judge against mise in Fastlane.

Sorry I may stop here. Maybe opening a new discussion or issue is a good idea? Or continue here. Now it seems that as I guessed before, [asdf-fastlane](https://github.com/mollyIV/asdf-fastlane) is correct and usable as a mise plugin, and now the problem lies in fastlane.

Thanks again for your contribution!

**Reply by lacostej** [reply]
The `fastlane` binary can be installed in multiple ways, and we have to be careful to keep those working.

When installed via bundler, if you run it via `bundle exec`, bundler will set the correct environment for you.

We can add support for mise. Note that I still think that for repeatable builds and managing dependencies & plugins, I still think that you should use a project specific `Gemfile`, and use bundler to install fastlane and its dependencies.

With regard to your issue, Cocoapods detects when it is ran standalone here
https://github.com/CocoaPods/CocoaPods/blob/3e7e5e87ab3d5d53728032cc158da39f80e818c1/bin/pod#L24C39-L28

We could probably do the same as long as we do not break other contexts.

One drawback is that your `mise` integration will only work *after* we've fixed fastlane to support this, i.e. you won't be able to run older versions through it.

Finally, maybe consider adjusting the environment printer here to display when fastlane is ran via `mise` ? https://github.com/fastlane/fastlane/blob/e1ab951b63aeb7a669b415708ffbe0e1346c0f59/fastlane/lib/fastlane/environment_printer.rb#L209-L221

**Reply by mollyIV** [reply]
With the help of the amazing folks from the Fastlane team, especially @AliSoftware, we figured out another way to make the Mise Fastlane plugin work. We leveraged [binstubs](https://github.com/rbenv/rbenv/wiki/Understanding-binstubs).

@rakuyoMo, could you please uninstall the existing `asdf-fastlane` plugin, install it again, and take it for a spin?

```bash
$ mise plugins uninstall fastlane -p
$ mise install 
```

One caveat: you must reinstall the `asdf-fastlane` plugin each time you change your Ruby version.

**Reply by rakuyoMo** [reply]
> With the help of the amazing folks from the Fastlane team, especially @AliSoftware, we figured out another way to make the Mise Fastlane plugin work. We leveraged [binstubs](https://github.com/rbenv/rbenv/wiki/Understanding-binstubs).

This is such exciting and good news! Thank you to everyone who has given so much!


> @rakuyoMo, could you please uninstall the existing `asdf-fastlane` plugin, install it again, and take it for a spin?

I will try it in 20 hours.

**Reply by rakuyoMo** [reply]
@mollyIV Something went wrong

```zsh
✗ mise install
mise plugin:fastlane ✓ git@github.com:mollyIV/asdf-fastlane.git#e6730ef
mise fastlane@2.222.0 ✓ installed

✗ mise exec -- fastlane --help
Could not find fastlane-2.222.0, CFPropertyList-3.0.7, addressable-2.8.7, artifactory-3.0.17, aws-sdk-s3-1.157.0, babosa-1.0.4, colored-1.2, commander-4.6.0, dotenv-2.8.1, emoji_regex-3.2.3, excon-0.111.0, faraday-1.10.3, faraday-cookie_jar-0.0.7, faraday_middleware-1.2.0, fastimage-2.3.1, gh_inspector-1.1.3, google-apis-androidpublisher_v3-0.54.0, google-apis-playcustomapp_v1-0.13.0, google-cloud-env-1.6.0, google-cloud-storage-1.47.0, highline-2.0.3, http-cookie-1.0.6, json-2.7.2, jwt-2.8.2, mini_magick-4.13.2, multipart-post-2.4.1, naturally-2.2.1, optparse-0.5.0, plist-3.7.1, rubyzip-2.3.2, security-0.1.5, simctl-1.6.10, terminal-notifier-2.0.0, terminal-table-3.0.2, tty-screen-0.8.2, tty-spinner-0.9.3, word_wrap-1.0.0, xcodeproj-1.25.0, xcpretty-0.3.0, xcpretty-travis-formatter-1.0.1, nkf-0.2.0, rexml-3.3.5, public_suffix-5.1.1, aws-sdk-core-3.201.4, aws-sdk-kms-1.88.0, aws-sigv4-1.9.1, faraday-em_http-1.0.0, faraday-em_synchrony-1.0.0, faraday-excon-1.1.0, faraday-httpclient-1.0.1, faraday-multipart-1.0.4, faraday-net_http-1.0.2, faraday-net_http_persistent-1.2.0, faraday-patron-1.0.0, faraday-rack-1.0.0, faraday-retry-1.0.3, google-apis-core-0.11.3, digest-crc-0.6.5, google-apis-iamcredentials_v1-0.17.0, google-apis-storage_v1-0.31.0, google-cloud-core-1.7.1, googleauth-1.8.1, mini_mime-1.1.5, domain_name-0.6.20240107, unicode-display_width-2.5.0, tty-cursor-0.7.1, atomos-0.1.3, claide-1.1.0, colored2-3.1.2, nanaimo-0.3.0, rouge-2.0.7, strscan-3.1.0, aws-eventstream-1.3.0, aws-partitions-1.963.0, jmespath-1.6.2, httpclient-2.8.3, representable-3.2.0, retriable-3.1.2, rake-13.2.1, google-cloud-errors-1.4.0, multi_json-1.15.0, os-1.1.4, signet-0.19.0, declarative-0.0.20, trailblazer-option-0.1.2, uber-0.1.0 in locally installed gems
Run `bundle install --gemfile /Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/Gemfile` to install missing gems.
```

But after I executed `bundle install --gemfile /Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/Gemfile`, executing `mise exec -- fastlane --help` again still reports this error.

```zsh
✗ bundle install --gemfile /Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/Gemfile
Using rake 13.2.1
Using base64 0.2.0
Using nkf 0.2.0
Using strscan 3.1.0
Using rexml 3.3.5
Using CFPropertyList 3.0.7
Using public_suffix 5.1.1
Using addressable 2.8.7
Using artifactory 3.0.17
Using atomos 0.1.3
Using aws-eventstream 1.3.0
Using aws-partitions 1.963.0
Using aws-sigv4 1.9.1
Using jmespath 1.6.2
Using aws-sdk-core 3.201.4
Using aws-sdk-kms 1.88.0
Using aws-sdk-s3 1.157.0
Using babosa 1.0.4
Using bundler 2.1.4
Using claide 1.1.0
Using colored 1.2
Using colored2 3.1.2
Using highline 2.0.3
Using commander 4.6.0
Using declarative 0.0.20
Using digest-crc 0.6.5
Using domain_name 0.6.20240107
Using dotenv 2.8.1
Using emoji_regex 3.2.3
Using excon 0.111.0
Using faraday-em_http 1.0.0
Using faraday-em_synchrony 1.0.0
Using faraday-excon 1.1.0
Using faraday-httpclient 1.0.1
Using multipart-post 2.4.1
Using faraday-multipart 1.0.4
Using faraday-net_http 1.0.2
Using faraday-net_http_persistent 1.2.0
Using faraday-patron 1.0.0
Using faraday-rack 1.0.0
Using faraday-retry 1.0.3
Using ruby2_keywords 0.0.5
Using faraday 1.10.3
Using http-cookie 1.0.6
Using faraday-cookie_jar 0.0.7
Using faraday_middleware 1.2.0
Using fastimage 2.3.1
Using gh_inspector 1.1.3
Using jwt 2.8.2
Using multi_json 1.15.0
Using os 1.1.4
Using signet 0.19.0
Using googleauth 1.8.1
Using httpclient 2.8.3
Using mini_mime 1.1.5
Using trailblazer-option 0.1.2
Using uber 0.1.0
Using representable 3.2.0
Using retriable 3.1.2
Using google-apis-core 0.11.3
Using google-apis-androidpublisher_v3 0.54.0
Using google-apis-playcustomapp_v1 0.13.0
Using google-cloud-env 1.6.0
Using google-apis-iamcredentials_v1 0.17.0
Using google-apis-storage_v1 0.31.0
Using google-cloud-errors 1.4.0
Using google-cloud-core 1.7.1
Using google-cloud-storage 1.47.0
Using json 2.7.2
Using mini_magick 4.13.2
Using naturally 2.2.1
Using optparse 0.5.0
Using plist 3.7.1
Using rubyzip 2.3.2
Using security 0.1.5
Using simctl 1.6.10
Using terminal-notifier 2.0.0
Using unicode-display_width 2.5.0
Using terminal-table 3.0.2
Using tty-screen 0.8.2
Using tty-cursor 0.7.1
Using tty-spinner 0.9.3
Using word_wrap 1.0.0
Using nanaimo 0.3.0
Using xcodeproj 1.25.0
Using rouge 2.0.7
Using xcpretty 0.3.0
Using xcpretty-travis-formatter 1.0.1
Using fastlane 2.222.0
Bundle complete! 1 Gemfile dependency, 89 gems now installed.
Bundled gems are installed into `../../../../../.local/share/mise/installs/fastlane/2.222.0/vendor/bundle`

✗ mise exec -- fastlane --help
Could not find fastlane-2.222.0, CFPropertyList-3.0.7, addressable-2.8.7, artifactory-3.0.17, aws-sdk-s3-1.157.0, babosa-1.0.4, colored-1.2, commander-4.6.0, dotenv-2.8.1, emoji_regex-3.2.3, excon-0.111.0, faraday-1.10.3, faraday-cookie_jar-0.0.7, faraday_middleware-1.2.0, fastimage-2.3.1, gh_inspector-1.1.3, google-apis-androidpublisher_v3-0.54.0, google-apis-playcustomapp_v1-0.13.0, google-cloud-env-1.6.0, google-cloud-storage-1.47.0, highline-2.0.3, http-cookie-1.0.6, json-2.7.2, jwt-2.8.2, mini_magick-4.13.2, multipart-post-2.4.1, naturally-2.2.1, optparse-0.5.0, plist-3.7.1, rubyzip-2.3.2, security-0.1.5, simctl-1.6.10, terminal-notifier-2.0.0, terminal-table-3.0.2, tty-screen-0.8.2, tty-spinner-0.9.3, word_wrap-1.0.0, xcodeproj-1.25.0, xcpretty-0.3.0, xcpretty-travis-formatter-1.0.1, nkf-0.2.0, rexml-3.3.5, public_suffix-5.1.1, aws-sdk-core-3.201.4, aws-sdk-kms-1.88.0, aws-sigv4-1.9.1, faraday-em_http-1.0.0, faraday-em_synchrony-1.0.0, faraday-excon-1.1.0, faraday-httpclient-1.0.1, faraday-multipart-1.0.4, faraday-net_http-1.0.2, faraday-net_http_persistent-1.2.0, faraday-patron-1.0.0, faraday-rack-1.0.0, faraday-retry-1.0.3, google-apis-core-0.11.3, digest-crc-0.6.5, google-apis-iamcredentials_v1-0.17.0, google-apis-storage_v1-0.31.0, google-cloud-core-1.7.1, googleauth-1.8.1, mini_mime-1.1.5, domain_name-0.6.20240107, unicode-display_width-2.5.0, tty-cursor-0.7.1, atomos-0.1.3, claide-1.1.0, colored2-3.1.2, nanaimo-0.3.0, rouge-2.0.7, strscan-3.1.0, aws-eventstream-1.3.0, aws-partitions-1.963.0, jmespath-1.6.2, httpclient-2.8.3, representable-3.2.0, retriable-3.1.2, rake-13.2.1, google-cloud-errors-1.4.0, multi_json-1.15.0, os-1.1.4, signet-0.19.0, declarative-0.0.20, trailblazer-option-0.1.2, uber-0.1.0 in locally installed gems
Run `bundle install --gemfile /Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/Gemfile` to install missing gems.
```

my `.mise.toml`:
```toml
[plugins]
swift-style-guide = 'https://github.com/RakuyoKit/asdf-swift-style-guide.git#archive/1.2.2'
fastlane = 'git@github.com:mollyIV/asdf-fastlane.git'

[tools]
pre-commit = "3.7.0"
swift-style-guide = "1.3.1"
swiftformat = "0.54.2"
swiftlint = "0.55.1"
tuist = "4.21.2"
fastlane = '2.222.0'
ruby = '3.3.0'
```

**Reply by mollyIV** [reply]
Hmm, did you uninstall the plugin first?

**Reply by rakuyoMo** [reply]
> Hmm, did you uninstall the plugin first?

Yes, I uninstalled the plugin and ruby ​​and reinstalled. I also executed `mise cache clean`.

----

**EDIT:**
I thought it was a problem with the order of installation, but after another test, it turned out not to be.

I uninstalled both ruby ​​and the plugin, then ran `mise install`. Ruby installs much slower than the plugin, so the plugin installs before ruby.

I tried keeping ruby, clearing the mise cache, uninstalling the plugin, and reinstalling the plugin. This time the plugin installed much slower, but I still got the same error.

----

**EDIT 2:**
By the way, I attach my mise version
```shell
✗ mise -v
2024.5.21 macos-arm64 (ef24c46 2024-05-23)
```

----

**EDIT 3:**
Even if I upgrade mise to `2024.8.5 macos-arm64 (98c44b8 2024-08-03)`, it still reports an error.

**Reply by mollyIV** [reply]
> Ruby installs much slower than the plugin, so the plugin installs before ruby.

There's a problem when you're installing Ruby and the plugin at the same time. The fastlane plugin requires some gems to work, and at the time of installing the plugin, your Ruby version must match the Ruby version you require in the mise configuration file. 

> I tried keeping ruby, clearing the mise cache, uninstalling the plugin, and reinstalling the plugin. This time the plugin installed much slower, but I still got the same error.

That's really surprising. It supposed to fix the issue. 

I tested it locally, and was able to reproduce the issue when installing both Ruby and the plugin at the same time. However, then uninstalling and installing again the plugin only did the trick. 

Can you please try again and show a few path outputs: 

<img width="738" alt="image" src="https://github.com/user-attachments/assets/98f53ccd-bbde-412b-b28a-190715e9995c">

**Reply by rakuyoMo** [reply]
```shell
➜ demo ✗ which ruby
/Users/rakuyo/.rvm/rubies/ruby-2.7.4/bin/ruby
➜ demo ✗ ruby -v
ruby 2.7.4p191 (2021-07-07 revision a21a3b7d23) [arm64-darwin20]
➜ demo ✗ which fastlane
/Users/rakuyo/.rvm/gems/ruby-2.7.4/bin/fastlane

➜ demo ✗ mise exec -- which ruby
/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/bin/ruby
➜ demo ✗ mise exec -- ruby -v
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [arm64-darwin23]
➜ demo ✗ mise exec -- which fastlane
/Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/bin/fastlane
```

**Reply by mollyIV** [reply]
And do you have a path like:

```
/Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/vendor/bundle/ruby/2.7.4/gems
```
?

I'd recommend Ruby >= 3.0.

**Reply by rakuyoMo** [reply]
I have this path: `/Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/vendor/bundle/ruby/2.7.0/gems`. There are many dependencies installed in the `gems` folder.

Judging from the path, fastlane installed via mise does not use the ruby ​​version declared in `.mise.toml`, even though the declared ruby ​​version was already installed when fastlane was installed.

**Reply by rakuyoMo** [reply]
Logically, when installing fastlane through mise, the ruby ​​version declared in `.mise.toml`, i.e. `mise exec -- which ruby`, should be used. This path may be inconsistent with `which ruby`.

Is this the problem now? If so, is there any way to solve this problem?

**Reply by mollyIV** [reply]
There may be an issue with your [mise local setup](https://mise.jdx.dev/getting-started.html). Please call `mise doctor` and resolve detected issues if any. 

I'd also make sure you have this: 

> "~/.local/bin" does not need to be in PATH. mise will automatically add its own directory to PATH when activated.

You can also check how your PATH looks like and if there's the Ruby version you expect it to be there.

**Reply by rakuyoMo** [reply]
I may have found the problem, my `PATH` is as follows. It starts with `ruby-2.7.4`.

```shell
➜ demo ✗ echo $PATH
/Users/rakuyo/.rvm/gems/ruby-2.7.4/bin:/Users/rakuyo/.rvm/gems/ruby-2.7.4@global/bin:/Users/rakuyo/.rvm/rubies/ruby-2.7.4/bin:/Users/rakuyo/oh-command-line-tools/bin:/opt/homebrew/Cellar/go@1.19/1.19.13/bin:/opt/homebrew/opt/python@3.10/bin:/Users/rakuyo/.local/share/mise/installs/pre-commit/3.7.0/bin:/Users/rakuyo/.local/share/mise/installs/swift-style-guide/1.3.1/bin:/Users/rakuyo/.local/share/mise/installs/swiftformat/0.54.2/bin:/Users/rakuyo/.local/share/mise/installs/swiftlint/0.55.1/bin:/Users/rakuyo/.local/share/mise/installs/tuist/4.21.2/bin:/Users/rakuyo/.local/share/mise/installs/fastlane/2.222.0/bin:/Users/rakuyo/.local/share/mise/installs/ruby/3.3.0/bin:/Users/rakuyo/.local/bin:/Library/Frameworks/Python.framework/Versions/3.10/bin:/opt/homebrew/bin:/opt/homebrew/sbin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/Library/Apple/usr/bin:/Users/rakuyo/.cargo/bin:/Applications/iTerm.app/Contents/Resources/utilities:/Users/rakuyo/.rvm/bin
```

When I executed `unset PATH && export PATH=...` in the terminal, removed the content related to `ruby-2.7.4`, and then reinstalled the plugin, it was able to install with the correct ruby ​​version, and `mise exec -- fastlane -v` also output the correct content.

The parsed content may not be relevant to the problem itself: I went through all the `rc` files I could find, commented out the content related to `.rvm`, but I still couldn't remove `ruby-2.7.4` from `PATH`.

In addition, I am not sure whether removing `ruby-2.7.4` from `PATH` is the right thing to do, it may affect those applications that use the global ruby ​​version?

--------

**UPDATE**:
I eventually found what was left from `~/.zlogin`, commented out the bit that loaded RVM, and now there's no ruby-2.7.4 in my `PATH` .

But my concerns above still exist.

If there are multiple ruby ​​versions in `PATH`, is there a way for the fastlane plugin to find the one specified in the `.mise.toml` file?

**Reply by mollyIV** [reply]
I'm glad you found the root cause. 

> If there are multiple ruby ​​versions in PATH, is there a way for the fastlane plugin to find the one specified in the .mise.toml file?

The fastlane plugin doesn't pick up the Ruby version itself. It's the mise setup that determines what Ruby version will be used to download the gems for. 

If there are multiple Ruby versions in PATH, the first one takes the precedence. 

Can we wrap up and confirm that the fastlane mise plugin works as expected then?

**Reply by rakuyoMo** [reply]
> Can we wrap up and confirm that the fastlane mise plugin works as expected then?

fastlane mise works as expected. Thanks for your work and dedication!

> If there are multiple Ruby versions in PATH, the first one takes the precedence.

I found that mise only imports mise-related paths in the directory containing the `.mise.toml` file, but rvm imports them globally.

This may be a problem for developers who use both mise and bundle to manage dependencies for different projects.

For example, A uses bundle to manage dependencies, which includes ruby ​​and fastlane, so `source "$HOME/.rvm/scripts/rvm"` is required in the rc file. Project B uses mise to manage dependencies. At this time, the ruby ​​version introduced in `PATH` by the above rvm command will interfere with the installation of fastlane.

Do you have any good suggestions for this?

**Reply by rakuyoMo** [reply]
This is probably a mise issue, I'm going to turn this over to ask for help.
